### DISCUSSION TITLE
Next.js Error with Dynamic Pages FUNCTION_INVOCATION_TIMEOUT

### DISCUSSION THREAD

**Discussion by abdulrahimiliasu** [vercel/next.js#32686] [Help]
I have a page in my Next.js site (`/flats/[id].js`) in `pages` folder.

Here is my page
```javascript
import React, { useState, useEffect } from "react";
import styled from "styled-components";

export default function FlatPage() {
  return (
    <Wrapper>
      <>Hello</>
    </Wrapper>
  );
}

const Wrapper = styled.div`
  padding: 20px;
`;
```

For some reason I get this error when navigating to the page.

    This Serverless Function has timed out.
    
    Your connection is working correctly.
    
    Vercel is working correctly.
    
    504: GATEWAY_TIMEOUT
    
    Code: FUNCTION_INVOCATION_TIMEOUT
    
    ID: fra1::kzjbt-1639939593230-a5edd370b414

CONSOLE LOGS:
```
[GET] /flats/16

15:35:23:41

Function Status: None

Edge Status: 504

Duration: 10009.74 ms

Init Duration: N/A

Memory Used: 243 MB

ID: ht6mw-1640010923026-45feee3fd9fb

User Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 15_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Mobile/15E148 Safari/604.1

2021-12-20T14:35:33.481Z 604f5a2d-3856-42c8-8a7c-72f1cc7d2b64 Task timed out after 10.01 seconds



[GET] /_next/data/qRevoA7v7ps_tbqeAEfjl/en/flats/16.json?id=16

15:35:11:99

Function Status: None

Edge Status: 504

Duration: 10007.67 ms

Init Duration: 397.63 ms
Memory Used: 284 MB
ID: ht6mw-1640010911602-b24295905176
User Agent: Mozilla/5.0 (iPhone; CPU iPhone OS 15_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Mobile/15E148 Safari/604.1

2021-12-20T14:35:22.853Z 9bdd796b-22a9-4159-b97a-6c9392189986 Task timed out after 10.01 seconds
```

I have no idea why the function timed out (nothing much on the page), plus it works fine locally too.

**Conditions for Serverless functions timeout From Next.js Docs**

> Timeout conditions checklist If you are seeing an execution timeout
> error, check the following possible causes:
> 
> The function isn't returning a response: 
> 
>  1. The function must return an HTTP response, even if that response is
>     an error. If no response is returned, the function will time out.
>  2. The function is taking too long to process a request: Check that any
>     API, or database requests you make in your function, are responding
>     within the "Serverless Function Execution Timeout (Seconds)" limit 
>     applicable to your plan.
>  3. You have an infinite loop within your function: Check that your
>     function is not making an infinite loop at any stage of execution.
>  4. There are upstream errors: Check that any external API or database
>     that you are attempting to call doesn't have any errors.

**Comment by aliabdy** [comment]


**Reply by abdulrahimiliasu** [reply]
Thanks, I already moved my production to Netlify. Everything works fine there

**Comment by PupoSDC** [comment]
Hey hey, I just wen through this problem. Maybe have a look here: https://github.com/vercel/next.js/issues/32745

**Reply by abdulrahimiliasu** [reply]
Sure maybe it has to do i18n, because I'm currently using it in my app. But it's really difficult to debug as there are no logs or trace of what the problem is. In the meantime I will move to Netlify. Thanks

**Reply by PupoSDC** [reply]
I understand your pain :| took me some 3 hours to figure it out, and I was only able to do it because my App is currently very small and I could pretty much start from scratch, adding stuff one by one until it broke. 

Good luck!

**Comment by ramirezsandin** [comment]
For any chance are you using **Material-UI**? and do you use a [shared layout](https://nextjs.org/docs/basic-features/layouts) in your app that has Material-UI components inside?, I've verified that with just these two conditions, the page returns a `504: GATEWAY_TIMEOUT` error if it has to render something server side:

- getServerSideProps: it doesn't matter if it's a dynamic page or not.
- getStaticProps with fallback:true or fallback:'blocking' and the page was not generated at build time.
- it doesn't matter if you're using i18n or not, it's the same.

**Reply by abdulrahimiliasu** [reply]
Yeah you right, I am using Material UI and a shared layout with Material UI, This only happens with **Vercel** though. Okay on Netlify.

**Reply by olhapi** [reply]
had the same issue just yesterday, checked today and it's working fine (nothing changed in codebase) 😆
and then it comes back after redeploy 🤯

**Reply by grichardomi** [reply]
What to do if you have  Material UI and a shared layout with Material UI?

**Reply by olhapi** [reply]
not sure if that'll help somebody, but in my case the issue was with material-ui icons, I've tried to replace them from another icons lib and that solved the issue (kind of)

**Reply by ghost** [reply]
> not sure if that'll help somebody, but in my case the issue was with material-ui icons, I've tried to replace them from another icons lib and that solved the issue (kind of)

Thank you, it works! I used Intellisense of VSCode and the auto completion and importation seemed a little problematic. After I used the [officially provided](https://mui.com/material-ui/material-icons/) material icon importation code by MUI, it was able to run normally.

**Reply by JustinShuttleworth** [reply]
@Antistable Can confirm! It was the material ui icon importation for me too. Struggled on this one for days and this comment here saved my app. Cheers.

**Reply by swax** [reply]
Same, it was the MUI icons causing Vercel to timeout. Thanks @olhapi !

**Reply by talha410900** [reply]
shukira

**Comment by aliabdy** [comment]
I solved this issue by changing default import to named import for importing mui icons.

https://mui.com/material-ui/guides/minimizing-bundle-size/

**Comment by alexthemark** [comment]
I ran into this yesterday with mui icons. I'm using typescript, and my issues were fixed by changing my import from `import { ExpandLess } from "@mui/icons-material";` to `import ExpandLess from "@mui/icons-material/ExpandLess";`

**Comment by Hassam790** [comment]
Hi,I am get the same error for dynamic pages on vercel. I am using mui and typescript. no above solution worked for me as i am not using mui icons. Is any one found solution for this issue?

**Comment by desarrolloBREspana** [comment]
Hi
I had problems when revalidating pages and it was because I use useMediaQuery, you have to put noSsr: true.
this fixed my problem

**Comment by oliviertassinari** [comment]
It could be an issue with tree-shaking, the app deployed on Vercel should bundle a single icon.

**Comment by IsaiahByDayah** [comment]
For what it's worth, I was running into this same issue when using `"@mui/icons-material": "^5.2.5"` but updating to latest (`"@mui/icons-material": "^5.11.0"`) seems to have fixed things with no changes to my imports

I see they updated the icons to use ESM so maybe that's what it was? ¯\_(ツ)_/¯

**Reply by ASchwad** [reply]
Did not yield the same result for me. Unfortunately, I still have to use the particular import to circumvent the tree-shaking issue

**Reply by JanderSilv** [reply]
Even with the latest version of `icons-material`, I had to change from named imports to default.

**Comment by avp1598** [comment]
Hi, I was struggling with this for a while. I am not using any MUI icons, but I read a comment about minimising bundle size and checked my initial load JS of the page and it was 1.18 Mb. I reduced the bundle size through dynamic importing and cleanup and got it down to 680 Kb and it worked fine. So I believe large bundle sizes can also cause timeout!
